[AOT] Fix runtime warnings

JamesNK opened the pull request:
Addresses https://github.com/dotnet/aspnetcore/issues/45473

* Run AOT analysis on all assemblies with IsTrimmable. Exceptions
  * SpaProxy, it uses configuration, and isn't urgent for preview 1
  * Blazor/components which is being addressed by https://github.com/dotnet/aspnetcore/pull/45580
* Suppresses all warnings from DI/hosting. These can be removed when https://github.com/dotnet/runtime/pull/79425 lands
* Almost all make generic type/method usages can either be suppressed or warn for dynamic code. There is a place in hosting where a type could be a struct, but realistically I don't think anyone would ever use a struct there. I've made it throw a runtime error if `RuntimeFeatures.IsDynamicCodeSupported` is false.

amcasey committed:
Enable AOT warnings

amcasey committed:
Suppress AOT warning

amcasey committed:
Simplify annotations on PropertyHelper

amcasey committed:
Annotate ProblemDetailJsonConverters

amcasey committed:
Fix whitespace

amcasey committed:
Annotate RouteValueDictionary

amcasey committed:
Guess suppression for WebApplicationBuilder

amcasey committed:
Annotate ManagedAuthenticatedEncryptorFactory

amcasey committed:
Address some RouteValueDictionary fallout

amcasey committed:
Annotate ILEmitTrie*

amcasey committed:
Annotate RequestDelegateFactory

amcasey committed:
Annotate Http*JsonExtensions

amcasey committed:
Annotate ParameterBindingMethodCache and EndpointMetadataPopulator

amcasey committed:
Annotate FSharpAsync codepath

amcasey committed:
Annotate RouteEndpointDataSource

amcasey committed:
DefaultServiceProvider fallout

JamesNK committed:
Update AOT suppressions

JamesNK committed:
Clean up

JamesNK committed:
Update

JamesNK committed:
Update

vitek-karas commented on src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs:
```diff
@@ -111,7 +112,8 @@ private static class AlgorithmActivator
         /// </summary>
-        public static Func<T> CreateFactory<T>([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type implementation)
+        [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "MakeGenericType is safe to use because implementation is either a KeyedHashAlgorithm or SymmetricAlgorithm type.")]
+        public static Func<T> CreateFactory<T>([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type implementation) where T : class
```
Nit: This whole class seems like a really complex way to do what it does - return a Func. I don't know the original reasons, but now with the delegate cachine Roslyn does I would think that a simple lambda would work the same way.

JamesNK replied:
I don't know the reason either. It can be looked at in the future. I want to keep this PR focused.

eerhardt replied:
Should we open an issue to track looking at this in the future, so it doesn't get forgotten?

vitek-karas commented on src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -267,7 +271,7 @@ void ConfigureStartup(HostBuilderContext context, IServiceCollection services)
     }
 
-    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2006:UnrecognizedReflectionPattern", Justification = "We need to call a generic method on IHostBuilder.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Constructing untrimmed generic types and methods with reference types")]
```
Is this still needed - since it seems you moved all of the MakeGeneric... into a helper.

halter73 commented on src/Identity/Extensions.Core/src/IdentityBuilder.cs:
```diff
@@ -21,6 +22,7 @@ public class IdentityBuilder
     /// <param name="services">The <see cref="IServiceCollection"/> to attach to.</param>
     public IdentityBuilder(Type user, IServiceCollection services)
     {
+        Debug.Assert(user.IsClass);
```
I can new up my own IdentityBuilder with an arbitrary type with no validation, right? I know that might be weird, but this is public. It looks like `AddIdentityCore<TUser>` doesn't validate either.

If there's any way whatsoever to make a `Debug.Assert` fail without resorting to something like private reflection, it should not be an assert. It should be a normal runtime check that throws for value types.

JamesNK replied:
You're right. Fortunately, there are class constraints everywhere on the types that user and role types are used with, so this would always fail, just not at this point. I'll add runtime validation.

halter73 reviewed (approved):
This seems like an improvement to me. I do wonder if we should have stronger runtime checks for the container and identity stuff. Maybe we should be adding new `: class` generic constraints to some APIs.

Should we file a tracking issue to remove the pragmas once https://github.com/dotnet/runtime/pull/79425 is merged?

eerhardt commented on src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -342,6 +335,30 @@ private void UseStartup([DynamicallyAccessedMembers(StartupLinkerOptions.Accessi
+        {
+            var containerType = configureContainerBuilder.GetContainerType();
+
+            if (containerType.IsValueType && !RuntimeFeature.IsDynamicCodeSupported)
+            {
+                throw new InvalidOperationException("ValueType startup container isn't supported with AOT.");
```
I don't think this is a valid approach to avoid the warning. This breaks the principle that "if you don't get any warnings, your app will work the same before and after publishing AOT".

If someone was using this functionality, and published for AOT, they wouldn't get any warnings and their app would work successfully at F5/`dotnet run` time. But when they went to publish the app, it would be broken.

Note this is different than the approach taken in https://github.com/dotnet/runtime/pull/79425. The difference is that DI is using a "feature switch" which gets enabled when `PublishAot=true` is set in the .csproj. That approach means that the app will still fail at F5/`dotnet run` time, if the app was using the AOT-incompatible functionality.

I think the existing hosting APIs should just be marked `RequiresDynamicCode` instead of suppressing these warnings.

eerhardt replied:
The same applies for the StartupLoader code/suppression.

JamesNK replied:
Startup containers are very rarely used. And I doubt anyone has created a struct startup container. I don't want to annotate the hosting APIs as RequiresDynamicCode for this one weird edge case.

Good point about `dotnet run`/`dotnet publish` difference. Should we add a flag like https://github.com/dotnet/runtime/pull/79425 does that ASP.NET Core can use to validate AOT compatibility? Or should there be one flag that is shared by libraries/applications?

eerhardt replied:
@vitek-karas @MichalStrehovsky @jkotas - thoughts on converting the DI feature switch in https://github.com/dotnet/runtime/pull/79425 to be a general purpose “Verify AOT Compatibility” feature switch that can be used by other libraries? Like for instance, here in aspnet?

eerhardt replied:
Another option is to always throw an exception for value types, AOT or not, if we think this functionality isn’t used with value types.

JamesNK replied:
Yeah, that's an option. We can publish it as a breaking change.

IMO regardless of whether we use it here or not, I think a generalized runtime verify AOT compatibility flag has some use. I think it provides a pragmatic middle ground in gray situations where something is compatible 99% of the time. Right now we have to choose between suppress and pretend everything is ok, and warn for everyone.

jkotas replied:
> general purpose “Verify AOT Compatibility” feature switch

This sounds like https://github.com/dotnet/runtime/issues/39806

vitek-karas replied:
Having something like https://github.com/dotnet/runtime/issues/39806 would be nice.

You're then left with a choice:
* Just disable the problematic case always (in our case here, always throw and potentially add class constrains in right places) - breaking change
* Rely on https://github.com/dotnet/runtime/issues/39806 to basically trigger `IsDynamicCodeSupported` based on a feature switch for you. Downside is that such solution is not granular - and so it would not be possible to disable dynamic code everywhere, except this one case.
* Introduce a new feature switch - which is what https://github.com/dotnet/runtime/pull/79425 is doing. Similar to the above, but it adds granularity. Downside is the added complexity and "yet another feature switch"
* Mark the API as AOT incompatible via the attributes - and hopefully figure out a new API which can be used instead

I think which one to do is a function of the specific scenario. If it's something super rare, I would go with either outright breaking change or generic feature switch. If it's still rare, but "real", then a dedicated feature switch might be better.

eerhardt commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -100,7 +100,8 @@ public RouteValueDictionary()
     /// </remarks>
-    [RequiresUnreferencedCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")]
+    [RequiresUnreferencedCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")] // PropertyStorage
+    [RequiresDynamicCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")] // PropertyStorage
```
"which may be trimmed" doesn't make sense here. This attribute isn't for "unreferenced" code, it is for requiring "dynamic" code.

eerhardt commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -585,6 +586,9 @@ public bool TryGetValue(string key, out object? value)
         "We do not need to additionally produce an error in this method since it is shared by trimmer friendly code paths.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050",
+        Justification = "The constructor that would result in _propertyStorage being non-null is annotated with RequiresDynamicCodeAttribute. " +
+        "We do not need to additionally produce an error in this method since it is shared by trimmer friendly code paths.")]
```
```suggestion
        "We do not need to additionally produce an error in this method since it is shared by AOT friendly code paths.")]
```

eerhardt commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -622,6 +626,9 @@ private void EnsureCapacity(int capacity)
         "We do not need to additionally produce an error in this method since it is shared by trimmer friendly code paths.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050",
+        Justification = "The constructor that would result in _propertyStorage being non-null is annotated with RequiresDynamicCodeAttribute. " +
+        "We do not need to additionally produce an error in this method since it is shared by trimmer friendly code paths.")]
```
```suggestion
        "We do not need to additionally produce an error in this method since it is shared by AOT friendly code paths.")]
```

halter73 commented on src/Shared/ProblemDetails/HttpValidationProblemDetailsJsonConverter.cs:
```diff
@@ -30,7 +34,8 @@ public static HttpValidationProblemDetails ReadProblemDetails(ref Utf8JsonReader
             if (reader.ValueTextEquals(Errors.EncodedUtf8Bytes))
             {
-                var errors = DeserializeErrors(ref reader, options);
+                var context = new ErrorsJsonContext(options);
```
I just looked at the build errors. Considering this mutates the JsonSerializerOptions, should we be configuring this earlier at startup. I know @brunolins16 has had to consider similar issues working on #45405.

brunolins16 replied:
I understand that using an internal `JsonContext` will make it easier to users, since they won't need to be annoyed with a bunch of types that are used internally, however, since those converters are internal is not even possible to use (HttpValidation)ProblemDetails with JSON source gen https://github.com/dotnet/aspnetcore/issues/43236 I believe we should think more about how to make `ProblemDetails` serialization safer, in special those converters that in my opinion are not really needed and are the source of the most of the issues.

brunolins16 replied:
I believe if the converters are not there, JSON would be able to source gen a `JsonTypeInfo<IDictionary<string, string[]>` only adding `HttpValidationProblemDetails` to the application `JsonContext`

brunolins16 replied:
I have a proposal here https://github.com/dotnet/aspnetcore/issues/44132#issuecomment-1256683443

JamesNK replied:
Working on this in https://github.com/dotnet/aspnetcore/pull/45646

eerhardt commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -793,6 +800,9 @@ public bool MoveNext()
             "We do not need to additionally produce an error in this method since it is shared by trimmer friendly code paths.")]
+        [UnconditionalSuppressMessage("AOT", "IL3050",
+            Justification = "The constructor that would result in _propertyStorage being non-null is annotated with RequiresDynamicCodeAttribute. " +
+            "We do not need to additionally produce an error in this method since it is shared by trimmer friendly code paths.")]
```
```suggestion
            "We do not need to additionally produce an error in this method since it is shared by AOT friendly code paths.")]
```

eerhardt commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -100,7 +100,8 @@ public RouteValueDictionary()
     /// Only public instance non-index properties are considered.
     /// </remarks>
-    [RequiresUnreferencedCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")]
+    [RequiresUnreferencedCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")] // PropertyStorage
+    [RequiresDynamicCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")] // PropertyStorage
```
What do the `PropertyStorage` comments at the end of the line mean?

JamesNK replied:
It's an indication of what is causing this warning in the method.

eerhardt commented on src/Shared/ProblemDetails/ProblemDetailsJsonConverter.cs:
```diff
@@ -8,15 +8,18 @@
-    [UnconditionalSuppressMessage("Trimmer", "IL2026", Justification = "Trimmer does not allow annotating overriden methods with annotations different from the ones in base type.")]
+    [RequiresUnreferencedCode("JSON serialization and deserialization of ProblemDetails.Extensions might require types that cannot be statically analyzed.")]
+    [RequiresDynamicCode("JSON serialization and deserialization of ProblemDetails.Extensions might require types that cannot be statically analyzed.")]
+    [UnconditionalSuppressMessage("Trimming", "IL2046:'RequiresUnreferencedCodeAttribute' annotations must match across all interface implementations or overrides.", Justification = "<Pending>")]
+    [UnconditionalSuppressMessage("AOT", "IL3051:'RequiresDynamicCodeAttribute' annotations must match across all interface implementations or overrides.", Justification = "<Pending>")]
```
Normally, the way we handle this is we mark the constructor of the type, or the whole type itself, as `RequiresUnreferencedCode`/`RequiresDynamicCode`. That says "creating one of these is not compatible", and moves the warning to the caller that instantiates it.

JamesNK replied:
The problem with adding the attributes to the entire type is that the converter type is then used in System.Text.Json serialization generated code. There is no chance to suppress it there.

eerhardt replied:
Suppressing the warning is not correct then, if we can't guarantee this code works with trimming and AOT. We will need to find another way designing this code to work with these form factors.

Can you open an issue for this, and add it as a comment here, so we can move forward with the rest of the changes?

JamesNK replied:
Working on this in https://github.com/dotnet/aspnetcore/pull/45646

eerhardt commented on src/Http/Http.Extensions/src/HttpRequestJsonExtensions.cs:
```diff
@@ -31,6 +31,7 @@ public static class HttpRequestJsonExtensions
     /// <param name="cancellationToken">A <see cref="CancellationToken"/> used to cancel the operation.</param>
     /// <returns>The task object representing the asynchronous operation.</returns>
     [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+    [RequiresDynamicCode(RequiresUnreferencedCodeMessage)]
```
This message is specific to "unreferenced" code, not "dynamic" code. Since this is public API, we should have a decent warning message about why this API requires "dynamic" code.

eerhardt commented on src/Http/Http.Extensions/src/HttpResponseJsonExtensions.cs:
```diff
@@ -29,6 +29,7 @@ public static partial class HttpResponseJsonExtensions
     /// <param name="cancellationToken">A <see cref="CancellationToken"/> used to cancel the operation.</param>
     /// <returns>The task object representing the asynchronous operation.</returns>
     [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+    [RequiresDynamicCode(RequiresUnreferencedCodeMessage)]
```
Same as above - these messages should be specific to "dynamic" code.

eerhardt replied:
Can someone remind me exactly what about RequestDelegateFactory that requires unreferenced and dynamic code? Is it only JSON (de)serialization? Or something else? it’s use of Linq.Expressions?

JamesNK replied:
* A LOT of reflection to get metadata about the endpoint
* JSON serialization/deserialization
* Expression trees
* MakeGenericType

eerhardt commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -28,11 +28,12 @@ namespace Microsoft.AspNetCore.Http;
+[UnconditionalSuppressMessage("Trimmer", "IL2072", Justification = "RequestDelegateFactory.Create has RequiresUnreferencedCode.")]
+[UnconditionalSuppressMessage("Trimmer", "IL2075", Justification = "RequestDelegateFactory.Create has RequiresUnreferencedCode.")]
+[UnconditionalSuppressMessage("Trimmer", "IL2077", Justification = "RequestDelegateFactory.Create has RequiresUnreferencedCode.")]
+[UnconditionalSuppressMessage("Trimmer", "IL3050", Justification = "RequestDelegateFactory.Create has RequiresDynamicCode.")]
```
Instead of suppressing all these warnings for the whole class, can we instead put `RequiresUnreferencedCode` and `RequiresDynamicCode` at the class level? That will turn off any warnings in the class, and mark all static methods as "RUC"/"RDC".

That way if we add a new static method, it won't be missed to add the correct attribute to it.

eerhardt commented on src/Http/Routing/src/Matching/JumpTableBuilder.cs:
```diff
@@ -85,11 +86,14 @@ public static JumpTable Build(int defaultDestination, int exitDestination, (stri
-        if (RuntimeFeature.IsDynamicCodeCompiled)
+        return MakeILEmitTrieJumpTableIfSupported(defaultDestination, exitDestination, pathEntries, fallback);
+
+        [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Guarded by IsDynamicCodeCompiled")]
```
Can you add a comment referencing https://github.com/dotnet/linker/issues/2715 here? That way this change can be reverted once that bug is fixed in the analyzer.

davidfowl replied:
We need a follow up bug for these things

eerhardt commented on src/Http/Routing/src/Patterns/DefaultRoutePatternTransformer.cs:
```diff
@@ -23,6 +23,8 @@ public DefaultRoutePatternTransformer(ParameterPolicyFactory policyFactory)
 
     [RequiresUnreferencedCode("This API may perform reflection on supplied parameter which may be trimmed if not referenced directly." +
         "Consider using a different overload to avoid this issue.")]
+    [RequiresDynamicCode("This API may perform reflection on supplied parameter which may be trimmed if not referenced directly." +
```
The message should be about why this uses "dynamic" code.

eerhardt commented on src/Http/Routing/src/Patterns/RoutePatternFactory.cs:
```diff
@@ -55,6 +55,7 @@ public static RoutePattern Parse([StringSyntax("Route")] string pattern)
     /// </param>
     /// <returns>The <see cref="RoutePattern"/>.</returns>
     [RequiresUnreferencedCode(RouteValueDictionaryTrimmerWarning.Warning)]
+    [RequiresDynamicCode(RouteValueDictionaryTrimmerWarning.Warning)]
```
These messages should be for "dynamic" code.

eerhardt commented on src/Http/Routing/src/Patterns/RoutePatternTransformer.cs:
```diff
@@ -34,6 +34,8 @@ public abstract class RoutePatternTransformer
     /// </remarks>
     [RequiresUnreferencedCode("This API may perform reflection on supplied parameter which may be trimmed if not referenced directly." +
         "Consider using a different overload to avoid this issue.")]
+    [RequiresDynamicCode("This API may perform reflection on supplied parameter which may be trimmed if not referenced directly." +
```
This message should be about dynamic code.

eerhardt commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -113,6 +113,8 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
     [UnconditionalSuppressMessage("Trimmer", "IL2026",
         Justification = "We surface a RequireUnreferencedCode in the call to the Map method adding this EndpointDataSource. The trimmer is unable to infer this.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050",
+        Justification = "We surface a RequiresDynamicCode in the call to the Map method adding this EndpointDataSource. The trimmer is unable to infer this.")]
```
Is this valid? I see a path that creates a `RouteEndpointDataSource` that isn't annotated.

https://github.com/dotnet/aspnetcore/blob/53124ab8cbf152f59120982f1c74e802e7970845/src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs#L145-L153

https://github.com/dotnet/aspnetcore/blob/53124ab8cbf152f59120982f1c74e802e7970845/src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs#L188-L199

https://github.com/dotnet/aspnetcore/blob/53124ab8cbf152f59120982f1c74e802e7970845/src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs#L410-L431

JamesNK replied:
This is fine. Map calls with `RequestDelegate` (instead of `Delegate`) are safe.

eerhardt replied:
We should update the justification then as it is not correct.

eerhardt replied:
Looking at the method, it looks like it will call into RequestDelegateFactory.Create even if it is a `RequestDelegate` when there are FilterFactories on the builder:

https://github.com/dotnet/aspnetcore/blob/53124ab8cbf152f59120982f1c74e802e7970845/src/Http/Routing/src/RouteEndpointDataSource.cs#L224-L230

In this case, are we sure it is still AOT safe?

JamesNK replied:
> Looking at the method, it looks like it will call into RequestDelegateFactory.Create even if it is a `RequestDelegate` when there are FilterFactories on the builder:
> 
> https://github.com/dotnet/aspnetcore/blob/53124ab8cbf152f59120982f1c74e802e7970845/src/Http/Routing/src/RouteEndpointDataSource.cs#L224-L230
> 
> In this case, are we sure it is still AOT safe?

Yes. The if statement will be false for `RequestDelegate` methods. TBH this type does too much.

JamesNK replied:
I updated the suppression justification to discuss `RequestDelegate` map methods.

eerhardt replied:
> The if statement will be false for RequestDelegate methods.

Even when there are FilterFactories? There’s an OR in that if statement.

eerhardt replied:
Can someone remind me what is not trim and AOT compatible about RequestDelegateFactory? Is it just the JSON (de)serialization? Or is there something else? The use of Linq.Expressions?

eerhardt commented on src/Identity/Core/src/IdentityBuilderExtensions.cs:
```diff
@@ -17,9 +18,12 @@ public static class IdentityBuilderExtensions
     public static IdentityBuilder AddDefaultTokenProviders(this IdentityBuilder builder)
     {
         var userType = builder.UserType;
+        Debug.Assert(userType.IsClass);
```
This is public API, so the Debug.Assert is basically useless. Anyone can create one of these IdentityBuilder types and pass in any Type for the UserType.

We should make this an exception that happens all the time (AOTing or not) instead in order to be correct.

eerhardt commented on src/Identity/Extensions.Core/src/IdentityBuilder.cs:
```diff
@@ -70,6 +72,7 @@ private IdentityBuilder AddScoped(Type serviceType, [DynamicallyAccessedMembers(
     /// </summary>
     /// <typeparam name="TValidator">The user validator type.</typeparam>
     /// <returns>The current <see cref="IdentityBuilder"/> instance.</returns>
+    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "MakeGenericType is safe because generic type and user type are reference types.")]
```
What does "generic type" in 

> because generic type and user type are reference types

mean? Should it just be

```suggestion
    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "MakeGenericType is safe because UserType is a reference type.")]
```

JamesNK committed:
PR feedback

JamesNK committed:
Clean up

JamesNK committed:
Apply suggestions from code review

Co-authored-by: Stephen Halter <halter73@gmail.com>
Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK commented:
I've reviewed all the PR comments and changes in this PR. I've simplified some annotations and removed some unnecessary ones.

Some items are still in discussion (startup container, attributes on problem details converters) and work todo (fix problem details errors serialization in converter). Everything else is resolved/improved.

JamesNK commented on src/Http/Routing/src/Builder/FallbackEndpointRouteBuilderExtensions.cs:
```diff
@@ -36,7 +36,6 @@ public static class FallbackEndpointRouteBuilderExtensions
     /// <c>{*path:nonfile}</c>. The order of the registered endpoint will be <c>int.MaxValue</c>.
     /// </para>
     /// </remarks>
-    [RequiresUnreferencedCode(EndpointRouteBuilderExtensions.MapEndpointTrimmerWarning)]
```
These were just wrong. We could backport removing this to .NET 7 if anyone raises them as an issue.

JamesNK committed:
Clean up

eerhardt commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -585,6 +585,9 @@ public bool TryGetValue(string key, out object? value)
     [UnconditionalSuppressMessage("Trimming", "IL2026",
         Justification = "The constructor that would result in _propertyStorage being non-null is annotated with RequiresUnreferencedCodeAttribute. " +
         "We do not need to additionally produce an error in this method since it is shared by trimmer friendly code paths.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050",
+        Justification = "The constructor that would result in _propertyStorage being non-null is annotated with RequiresDynamicCodeAttribute. " +
+        "We do not need to additionally produce an error in this method since it is shared by AOT friendly code paths.")]
```
These suppressions are no longer valid because the `public RouteValueDictionary(object? values)` constructor is no longer annotated with `RequiresDynamicCodeAttribute`.

JamesNK replied:
Some dynamic code was still used in `RouteValueDictionary(object? values)`, but I replaced it with some simple reflection is dynamic code isn't supported.

eerhardt commented on src/Identity/Extensions.Core/src/IdentityBuilder.cs:
```diff
@@ -21,6 +22,11 @@ public class IdentityBuilder
     {
+        if (user.IsValueType)
+        {
+            throw new ArgumentException("User type can't be a value type.", nameof(user));
```
Should these exception messages be in a .resx file?

JamesNK replied:
No, we don't worry about it in aspnetcore. Maybe we should, but that's a bigger discussion 😄

eerhardt commented on src/Shared/ProblemDetails/HttpValidationProblemDetailsJsonConverter.cs:
```diff
@@ -7,18 +7,22 @@
     }
 
-    [RequiresUnreferencedCode("JSON serialization and deserialization ProblemDetails.Extensions might require types that cannot be statically analyzed. ")]
+    [RequiresUnreferencedCode("JSON serialization and deserialization of ProblemDetails.Extensions might require types that cannot be statically analyzed.")]
+    [RequiresDynamicCode("JSON serialization and deserialization of ProblemDetails.Extensions might require types that cannot be statically analyzed.")]
```
Are these annotations still necessary now that we are using the JSON source generator?

JamesNK replied:
Working on this in https://github.com/dotnet/aspnetcore/pull/45646

eerhardt commented on src/Shared/ProblemDetails/HttpValidationProblemDetailsJsonConverter.cs:
```diff
@@ -30,7 +34,8 @@ public static HttpValidationProblemDetails ReadProblemDetails(ref Utf8JsonReader
             if (reader.ValueTextEquals(Errors.EncodedUtf8Bytes))
             {
-                var errors = DeserializeErrors(ref reader, options);
+                var context = new ErrorsJsonContext(options);
```
Do we need to create a new `ErrorsJsonContext` instance every time through this loop?

JamesNK replied:
Working on this in https://github.com/dotnet/aspnetcore/pull/45646

eerhardt commented on src/Shared/EndpointMetadataPopulator.cs:
```diff
@@ -12,6 +12,7 @@
 namespace Microsoft.AspNetCore.Http;
 
 [UnconditionalSuppressMessage("Trimmer", "IL2060", Justification = "Trimmer warnings are presented in RequestDelegateFactory.")]
+[UnconditionalSuppressMessage("AOT", "IL3050", Justification = "AOT warnings are presented in RequestDelegateFactory.")]
```
These shouldn't be suppressed. Instead this class should be marked with `RequiresUnreferenecedCode` and `RequiresDynamicCode`. That way if any new code starts calling it, they will get warnings as well.

eerhardt replied:
Looks like this is also called from

https://github.com/dotnet/aspnetcore/blob/a8813d4bcbf2f940cd3e159d266fdd039a390922/src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs#L370-L373

This is why suppressions like this aren't valid.

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK commented on src/Shared/ProblemDetails/ProblemDetailsJsonConverter.cs:
```diff
@@ -100,6 +107,7 @@ internal static bool TryReadStringProperty(ref Utf8JsonReader reader, JsonEncode
 
     [RequiresUnreferencedCode("JSON serialization and deserialization of ProblemDetails.Extensions might require types that cannot be statically analyzed.")]
+    [RequiresDynamicCode("JSON serialization and deserialization of ProblemDetails.Extensions might require types that cannot be statically analyzed.")]
     internal static void WriteProblemDetails(Utf8JsonWriter writer, ProblemDetails value, JsonSerializerOptions options)
```
Working on this in https://github.com/dotnet/aspnetcore/pull/45646

eerhardt commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -112,7 +112,11 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
-        Justification = "We surface a RequireUnreferencedCode in the call to the Map method adding this EndpointDataSource. The trimmer is unable to infer this.")]
+        Justification = "We surface a RequireUnreferencedCode in the call to the Map methods adding route handlers to this EndpointDataSource. Analysis is unable to infer this. " +
+        "Map methods that configure a RequestDelegate don't use trimmer unsafe features.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050",
+        Justification = "We surface a RequiresDynamicCode in the call to the Map methods adding route handlers this EndpointDataSource. Analysis is unable to infer this. " +
+        "Map methods that configure a RequestDelegate don't use AOT unsafe features.")]
```
This method is pretty big, and I can imagine people adding to it over time.

Instead of suppressing these warnings for the whole method, can we extract just the 2 calls to `RequestDelegateFactory` into local methods, and then just suppress there? That way if someone adds new code that is AOT/trimming unsafe, they will get warnings.

JamesNK replied:
I tried that, and the logic is too intermingled. I didn't feel comfortable refactoring something I'm not that familiar. Can log improving this as a follow up task.

eerhardt commented on src/Http/Routing/src/Builder/EndpointFilterExtensions.cs:
```diff
@@ -13,12 +13,15 @@ namespace Microsoft.AspNetCore.Http;
     /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
     /// <param name="filter">The <see cref="IEndpointFilter"/> to register.</param>
     /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the route handler.</returns>
+    [RequiresDynamicCode(FilterRequiresDynamicCodeWarning)]
     public static TBuilder AddEndpointFilter<TBuilder>(this TBuilder builder, IEndpointFilter filter) where TBuilder : IEndpointConventionBuilder =>
```
@captainsafia @davidfowl - Will these filter APIs work correctly in AOT when we have the minimal API source generator?

JamesNK replied:
The API to get the value today is something like `T GetValue<T>(...)`, accessed via `MakeGenericMethod`. Won't work with a struct parameter.

Adding a `object GetValue(...)` method might be a way to get AOT compatibility. There would be the cost of boxing, but better than an error. The alternative would be figuring out code gen.

MichalStrehovsky replied:
How important is avoiding the boxing? Would we be comfortable with using `object GetValue(...)` everywhere irrespective of the runtime? (I.e. box even if we had a JIT.)

If boxing wouldn't be acceptable elsewhere, would it be possible to extract the problematic pattern into a standalone console app so that we can zoom in on it separately and see if something could be done on the runtime side?

davidfowl replied:
Yes it will work with the source generator. We have a prototype. This needs to be removed.

eerhardt replied:
> This RequiresDynamicCode needs to be removed.

I believe then it means that the suppressions added here:

https://github.com/dotnet/aspnetcore/blob/9175347822be11e4aa1376e69a4c5f17db731388/src/Http/Routing/src/RouteEndpointDataSource.cs#L114-L122

need to be removed as well? Basically meaning even the `MapAction(RequestDelegate)` overloads need to be marked as `RequiresDynamicCode`. And only the source generator will enable AOT'ing minimal APIs successfully.

davidfowl replied:
> Basically meaning even the MapAction(RequestDelegate) overloads need to be marked as RequiresDynamicCode. And only the source generator will enable AOT'ing minimal APIs successfully.

No, this code needs to be refactored so that there are abstractions that flow into this code. The source generator will reuse this code with a source generated implementation. The Map(Delegate) calls will add the unsafe implementation and the Map(RequestDelegate) overloads will add a safe implementation.

JamesNK committed:
Update src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

JamesNK committed:
Update src/Hosting/Hosting/src/Internal/StartupLoader.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

JamesNK committed:
[AOT] Fixing the problems with ProblemDetails (#45646)

JamesNK committed:
PR feedback

davidfowl commented on src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs:
```diff
@@ -66,7 +71,22 @@ public ValueTask WriteAsync(ProblemDetailsContext context)
+    // Additional values are specified on JsonSerializerContext to support some values for extensions.
+    // For example, the DeveloperExceptionMiddleware serializes its complex type to JsonElement, which problem details then needs to serialize.
     [JsonSerializable(typeof(ProblemDetails))]
+    [JsonSerializable(typeof(JsonElement))]
+    [JsonSerializable(typeof(string))]
+    [JsonSerializable(typeof(decimal))]
+    [JsonSerializable(typeof(float))]
+    [JsonSerializable(typeof(double))]
+    [JsonSerializable(typeof(int))]
+    [JsonSerializable(typeof(long))]
+    [JsonSerializable(typeof(Guid))]
+    [JsonSerializable(typeof(Uri))]
+    [JsonSerializable(typeof(TimeSpan))]
+    [JsonSerializable(typeof(DateTime))]
+    [JsonSerializable(typeof(DateTimeOffset))]
```
We should undo this.

JamesNK replied:
This is useful because it allows common type values as extension values to just work, e.g. adding a string value to problem details extensions.

Other than a small size cost when problem details aren't trimmed away, this doesn't hurt. Are you sure you want it removed?

davidfowl replied:
I don't mind keeping it for now and removing it with @brunolins16's JSON changes, but I don't think this is generally useful.

JamesNK replied:
Ok! I'd like to keep as is in this PR (rather than reverting to what it was, which doesn't work), and we can iterate on it during .NET 8 as we learn more.

davidfowl commented on src/Http/Http.Extensions/src/HttpResponseJsonExtensions.cs:
```diff
@@ -18,6 +18,8 @@ public static partial class HttpResponseJsonExtensions
 {
     private const string RequiresUnreferencedCodeMessage = "JSON serialization and deserialization might require types that cannot be statically analyzed. " +
```
@brunolins16 this will all be undone with your changes right?

brunolins16 replied:
I don't believe, the unsafe API will still need to be attributed, however, RDF + MVC should not call them anymore.

davidfowl commented on src/Http/Routing/src/Builder/EndpointFilterExtensions.cs:
```diff
@@ -13,12 +13,15 @@ namespace Microsoft.AspNetCore.Http;
 /// </summary>
 public static class EndpointFilterExtensions
 {
+    internal const string FilterRequiresDynamicCodeWarning = "Filter factories generate dynamic code and aren't compatible with native AOT applications.";
```
This isn't true. Filters work fine.

eerhardt replied:
If you have filters, we will call into RequestDelegateFactory.Create, which is marked `RequiresDynamicCode`:

https://github.com/dotnet/aspnetcore/blob/9175347822be11e4aa1376e69a4c5f17db731388/src/Http/Routing/src/RouteEndpointDataSource.cs#L228-L235

JamesNK replied:
I looked into `RequestDelegateFactory` + filters, and there is logic for falling back to non-dynamic code if AOT:
* https://github.com/dotnet/aspnetcore/blob/589aa11b5c631ce719e0530d66be8324a6d79169/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L546-L551
* https://github.com/dotnet/aspnetcore/blob/589aa11b5c631ce719e0530d66be8324a6d79169/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L611-L619

I think we definitely need AOTed testing of this area. It would be very easy for someone to unknowingly break support.

davidfowl replied:
No I think in .NET 8, RequestDelegateFactory will be completely unsafe for AOT. That's what [this work item](https://github.com/dotnet/aspnetcore/issues/45524) is about.

JamesNK replied:
Map + RequestDelegate will still go through `RequestDelegateFactory`. I want to test that is ok. And test that it is ok with a filter

davidfowl replied:
> Map + RequestDelegate will still go through RequestDelegateFactory (unless that is rewritten).

It will be rewritten.

> I want to test that is ok. And test that Map + RequestDelegate + filter it is ok.

You're testing the current state of the code. Adding filters to an endpoint is not fundamentally unsafe and thus cannot be marked as such.

davidfowl commented on src/Http/Routing/src/Builder/EndpointFilterExtensions.cs:
```diff
@@ -29,6 +32,7 @@ public static class EndpointFilterExtensions
     /// <typeparam name="TFilterType">The type of the <see cref="IEndpointFilter"/> to register.</typeparam>
     /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
     /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the route handler.</returns>
+    [RequiresDynamicCode(FilterRequiresDynamicCodeWarning)]
```
These should all be undone.

JamesNK committed:
PR feedback

JamesNK committed:
Fix

JamesNK commented:
@eerhardt @davidfowl All feedback applied. Are there any other changes, or can this get approved and merged 🙏

eerhardt commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -100,7 +100,7 @@ public RouteValueDictionary()
     /// Only public instance non-index properties are considered.
     /// </remarks>
-    [RequiresUnreferencedCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")]
+    [RequiresUnreferencedCode("This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.")] // PropertyStorage
```
I don't understand what the added comment means. Was this change intentional?

eerhardt commented on src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs:
```diff
@@ -48,17 +50,20 @@ public bool CanWrite(ProblemDetailsContext context)
     [UnconditionalSuppressMessage("Trimming", "IL2026",
-        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed and we need to fallback" +
-        "to reflection-based. The ProblemDetailsConverter is marked as RequiresUnreferencedCode already.")]
+        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed. The property is annotated with a warning")]
+    [UnconditionalSuppressMessage("Trimming", "IL3050",
+        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed. The property is annotated with a warning")]
```
(nit)

```suggestion
        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed. The property is annotated with RequiresUnreferencedCode.")]
    [UnconditionalSuppressMessage("Trimming", "IL3050",
        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed. The property is annotated with RequiresDynamicCode.")]
```

eerhardt commented on src/Http/Routing/src/Matching/JumpTableBuilder.cs:
```diff
@@ -85,13 +86,15 @@ public static JumpTable Build(int defaultDestination, int exitDestination, (stri
 
         // Use the ILEmitTrieJumpTable if the IL is going to be compiled (not interpreted)
-        if (RuntimeFeature.IsDynamicCodeCompiled)
+        return MakeILEmitTrieJumpTableIfSupported(defaultDestination, exitDestination, pathEntries, fallback);
```
I think this change is unnecessary and can be reverted. The existing code should be able to pass CI.

JamesNK replied:
I'll leave it as is. It will be changed again when analysis supports IsDynamicCodeCompiled.

eerhardt commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -253,9 +251,31 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+            "Map methods that configure a RequestDelegate don't use trimmer unsafe features.")]
+        [UnconditionalSuppressMessage("AOT", "IL3050",
+            Justification = "We surface a RequiresDynamicCode in the call to the Map methods adding route handlers this EndpointDataSource. Analysis is unable to infer this. " +
+            "Map methods that configure a RequestDelegate don't use AOT unsafe features.")]
```
Is there some way we can verify this?

One idea is to add a "Native AOT"-test that uses filter factories and ensures they work correctly.

If it is possible to verify this, we can do it in a follow-up issue/PR. It doesn't need to be addressed in this PR.

JamesNK replied:
I'll create follow-up issues for areas to test with native AOT enabled. Any code that tests `IsDynamicCodeSupported` is a good candidate.

davidfowl replied:
The plan is to undo this anyways. This should be suppression free after we execute https://github.com/dotnet/aspnetcore/issues/45524. There will be new API changes that remove the use of RequestDelegateFactory.

eerhardt commented on src/Identity/Extensions.Core/src/IdentityBuilder.cs:
```diff
@@ -21,6 +22,11 @@ public class IdentityBuilder
     {
+        if (user.IsValueType)
+        {
+            throw new ArgumentException("User type can't be a value type.", nameof(user));
```
Should we add tests for passing a ValueType to these APIs? To ensure they break if someone removes the checks.

eerhardt commented on src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs:
```diff
@@ -328,3 +352,10 @@ private Task DisplayRuntimeException(HttpContext context, Exception ex)
     }
 }
+
+internal record ExceptionExtensionData(string Details, IHeaderDictionary Headers, string Path, string? Endpoint, RouteValueDictionary? RouteValues);
```
For my learning:

In dotnet/runtime we typically don't use `record` because of the bloat associated with it - C# generates members you don't use/need. Do we have guidelines in dotnet/aspnetcore when to use `record` and when not? Or don't we care as much?

JamesNK replied:
I think we don't care as much. It's internal, so if we crack down on records in aspnetcore (we have them in other places) then it's simple enough to change.

jkotas replied:
For reference, this is how much IL is generated by this single `record` line: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgH1kigBMACAUQTBgAcNsIA7KjGJgZ0aYBEBDDHwAUAAQCMABjI8YgvBzRkAkgAkYfEjCg9sYBsz5QAnmTUatCsuKkAFAQAtF1gPyUmJWhGxMMigEoQAK5sAGp8uIEwOnrchkauAcEwYREwHACUANwAsABQeXkiAExkiaHhkdH6THF5AN55AL4Fud5sUABmfDTKZprautW1uQ25jUA==

This `record` will translate to about 5kB of binary footprint in the native aot image. 80% of it is dynamically unreachable code, but it is impossible for the trimmer to prove that it is unreachable because it is all behind potentially reachable interfaces.

JamesNK replied:
Impressive! Looks like we should clean up record usage in aspnetcore.

eerhardt reviewed (approved):
This looks like a great step forward! Thanks for all the great work here.

I just had a few nits, a question, and a follow-up recommendation for the future.

This LGTM. Let's get this in and make more progress.

JamesNK committed:
Update src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

JamesNK committed:
PR feedback

davidfowl commented on src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs:
```diff
@@ -214,6 +213,31 @@ await _problemDetailsService.WriteAsync(new()
         }
     }
 
+    [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "Values set on ProblemDetails.Extensions are supported by the default writer.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Values set on ProblemDetails.Extensions are supported by the default writer.")]
+    private ProblemDetails CreateProblemDetails(ErrorContext errorContext, HttpContext httpContext)
```
Why not let it warn until we have the real fix?

davidfowl commented on src/Identity/Extensions.Core/src/IdentityBuilder.cs:
```diff
@@ -32,7 +38,14 @@ public IdentityBuilder(Type user, IServiceCollection services)
     public IdentityBuilder(Type user, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type role, IServiceCollection services) : this(user, services)
-        => RoleType = role;
+    {
+        if (role.IsValueType)
```
Is this a breaking change?

JamesNK replied:
`RoleType` and `UserType` are later used to generate types, all of which have a `where T : class` constraint.

Throwing here means the error is different and it is thrown in a different place. But a value type here never worked.

davidfowl reviewed (approved):
I'd like to see issues for:
- The developer exception page WRT what we're doing JSON wise there
- ProblemDetails and undoing what's there now 
- Undoing the DI changes when those other changes land
- Doing API work in the Map* handlers to remove suppressions in the RouteEndpointDataSource

JamesNK commented:
Follow-up issues:
* https://github.com/dotnet/aspnetcore/issues/45863
* https://github.com/dotnet/aspnetcore/issues/45860
* https://github.com/dotnet/aspnetcore/issues/45862
* https://github.com/dotnet/aspnetcore/issues/45863

JamesNK commented:
Merged! Thanks everyone for your feedback.
